### PR TITLE
crimson/onode-staged-tree: support empty ns and oid

### PR DESCRIPTION
**Pull request by cyx1231st**
Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h [hunk 1]
```diff
@@ -140,9 +140,9 @@ struct string_key_view_t {
   // presumably the maximum string length is 2KiB
   using string_size_t = uint16_t;
   static constexpr auto MAX = std::numeric_limits<string_size_t>::max();
-  static constexpr auto MIN = string_size_t(0u);
-  static auto is_valid_size(size_t size) {
-    return (size > MIN && size < MAX);
+  static constexpr auto MIN = MAX - 1;
+  static bool is_valid_size(size_t size) {
+    return size < MIN;
```
Comment (athanatos) [inline src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:145]: I don't really understand the role of MIN here.

**Review by athanatos — APPROVED** [review]
MIN and MAX seem misleading in string_key_view_t, but otherwise fine.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h [hunk 2]
```diff
@@ -147,8 +152,17 @@ class KVPool {
     std::random_device rd;
     for (unsigned i = range2.first; i < range2.second; ++i) {
       for (unsigned j = range1.first; j < range1.second; ++j) {
-        auto ns_size = (unsigned)str_sizes[rd() % str_sizes.size()];
-        auto oid_size = (unsigned)str_sizes[rd() % str_sizes.size()];
+        unsigned ns_size;
```
Comment (mattbenjamin) [inline src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h:155]: unsigned?  could this be uint32_t, please?

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:145]: You can assume `MIN` and `MAX` are invalid/impossible values now. They are currently reserved to implement ns/oid deduplication across onode tree nodes.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h:155]: Sure.
This file is for test/perf-tool purposes, so I didn't use types very carefully.

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:145]: Please also refer to the comments for `struct string_key_view_t`.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h]: fixed.

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h (hunk 1 above)
Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:145]: I mostly object to the fact that < MIN is "valid".

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:145]: OK, I'll add a `UPPER_BOUND` to be `MAX - 2`

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h]: see https://github.com/ceph/ceph/pull/39333/commits/54f87106649aafa1c6a0cf77e72eb04df3a5a41c

**Pull request merged**